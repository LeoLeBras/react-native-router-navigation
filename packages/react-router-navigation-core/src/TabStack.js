/* @flow */

import * as React from 'react'
import { matchPath, type RouterHistory } from 'react-router'
import invariant from 'invariant'
import HistoryUtils from './HistoryUtils'
import StackUtils from './StackUtils'
import RouteUtils from './RouteUtils'
import StateUtils from './StateUtils'
import type {
  NavigationState,
  TabsRendererProps,
  Tab,
  Route,
  HistoryRootIndex,
  HistoryNodes,
} from './TypeDefinitions'

type Props = {
  history: RouterHistory,
  children: React$Node[],
  render: (props: TabsRendererProps<>) => React$Node,
  lazy?: boolean,
}

type State = {|
  tabs: Tab[],
  navigationState: NavigationState<>,
  historyRootIndex: HistoryRootIndex,
  historyNodes: HistoryNodes,
|}

class TabStack extends React.Component<Props, State> {
  unlistenHistory: ?Function = null

  constructor(props: Props) {
    super(props)
    const { children, history } = props
    invariant(
      history,
      'The prop `history` is marked as required in `TabStack`, but its value is `undefined` in TabStack',
    )
    invariant(
      children || React.Children.count(children) > 0,
      'A <TabStack /> must have child elements',
    )
    const { index, location } = history
    const entries = history.entries || [location]
    const tabs = children && StackUtils.create(children, props)
    const navigationState = StateUtils.initialize(
      tabs,
      location,
      entries,
      'nodes',
    )
    invariant(
      navigationState.index !== -1,
      'There is no route defined for path « %s »',
      location.pathname,
    )
    const initialRoute = navigationState.routes[navigationState.index]
    const historyRootIndex = index
    const historyNodes = { [initialRoute.name]: entries.slice(index) }
    this.state = { tabs, navigationState, historyRootIndex, historyNodes }
  }

  componentDidMount() {
    const { history } = this.props
    this.unlistenHistory = HistoryUtils.listen(history, this.onHistoryChange)
  }

  componentWillUnmount() {
    if (this.unlistenHistory) this.unlistenHistory()
  }

  componentWillReceiveProps(nextProps: Props) {
    const { tabs } = this.state
    const { children: nextChildren, history } = nextProps
    const { location } = history
    const entries = history.entries || [location]
    const nextTabs = StackUtils.create(nextChildren, nextProps)
    const nextTab = nextTabs.find(tab => matchPath(location.pathname, tab))
    const nextRoute = nextTab && RouteUtils.create(nextTab, location)
    if (nextRoute && !StackUtils.shallowEqual(tabs, nextTabs)) {
      const nextNavigationState = StateUtils.initialize(
        nextTabs,
        location,
        entries,
        'nodes',
      )
      invariant(
        nextNavigationState.index !== -1,
        'There is no route defined for path « %s »',
        location.pathname,
      )
      this.setState({
        tabs: nextTabs,
        navigationState: nextNavigationState,
      })
    }
  }

  onHistoryChange = (history: RouterHistory, nextHistory: RouterHistory) => {
    const { location } = nextHistory
    const { navigationState, tabs } = this.state
    const { routes } = navigationState
    const currentRoute = routes[navigationState.index]
    const nextTab = tabs.find(tab => matchPath(location.pathname, tab))
    if (nextTab) {
      const staleRoute = routes.find(route => route.name === nextTab.path)
      const nextRoute = RouteUtils.create(nextTab, location, staleRoute)
      if (nextRoute && !RouteUtils.equal(currentRoute, nextRoute)) {
        this.setState(prevState => ({
          navigationState: StateUtils.changeIndex(
            prevState.navigationState,
            nextRoute,
          ),
        }))
      }
    }
  }

  onIndexChange = (arg: number | Route) => {
    const { history } = this.props
    const { tabs, navigationState, historyRootIndex } = this.state
    const index = StateUtils.getRouteIndex(navigationState, arg)
    const nextRoute = navigationState.routes[index]
    const nextTab = tabs.find(tab => tab.path === nextRoute.name)
    if (nextTab && index !== navigationState.index) {
      const location = HistoryUtils.createLocation(history, nextTab)
      this.setState(
        {
          navigationState: StateUtils.changeIndex(
            navigationState,
            RouteUtils.create(nextTab, location, nextRoute) || index,
          ),
        },
        () => {
          history.replace(location.pathname, location.state)
        },
      )
    } else {
      const n = historyRootIndex - history.index
      if (n < 0) {
        history.go(n)
      } else if (nextTab) {
        if (typeof nextTab.onReset === 'function') {
          nextTab.onReset()
        } else if (nextTab.initialPath) {
          history.replace(nextTab.initialPath)
        } else if (nextTab.path) {
          history.replace(nextTab.path)
        }
      }
    }
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return (
      this.state.tabs !== nextState.tabs ||
      this.state.navigationState !== nextState.navigationState
    )
  }

  renderTab = (route: Route) => {
    const { lazy } = this.props
    const { historyNodes } = this.state
    if (lazy && !historyNodes[route.name]) return null
    const children = React.Children.toArray(this.props.children)
    const child = children.find(({ props }) => props.path === route.name)
    if (!child) return null
    return React.cloneElement(child, route)
  }

  render() {
    return this.props.render({
      navigationState: this.state.navigationState,
      tabs: this.state.tabs,
      onIndexChange: this.onIndexChange,
      renderTab: this.renderTab,
    })
  }
}

export default TabStack
