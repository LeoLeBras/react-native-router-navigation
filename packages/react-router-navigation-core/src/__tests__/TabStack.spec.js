/* @noflow */

import * as React from 'react'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'

import './__mocks__'
import { componentFactory, renderTabView } from './utils'
import TabStack from './../TabStack'

describe('<TabStack />', () => {
  const IndexComponent = componentFactory('Index')
  const HelloComponent = componentFactory('Hello')
  const GoodbyeComponent = componentFactory('Goodbye')

  it('should render correctly', () => {
    const history = createHistory()
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack
              history={contextRouter.history}
              render={props => {
                expect(typeof props.onIndexChange).toBe('function')
                expect(typeof props.renderTab).toBe('function')
                expect(props.tabs).toMatchObject([
                  {
                    exact: true,
                    path: '/',
                    component: IndexComponent,
                  },
                  {
                    path: '/hello',
                    component: HelloComponent,
                  },
                  {
                    path: '/goodbye',
                    component: GoodbyeComponent,
                  },
                ])
                expect(props.navigationState).toMatchObject({
                  index: 0,
                  routes: [
                    {
                      key: '/',
                      routeName: '/',
                      routeMatch: {
                        url: '/',
                        isExact: true,
                        path: '/',
                        params: {},
                      },
                    },
                    {
                      key: '/hello',
                      routeName: '/hello',
                      routeMatch: null,
                    },
                    {
                      key: '/goodbye',
                      routeName: '/goodbye',
                      routeMatch: null,
                    },
                  ],
                })
                return renderTabView(props)
              }}
            >
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with initialIndex and initialEntries props ', () => {
    const history = createHistory({
      initialIndex: 3,
      initialEntries: ['/', '/yolo', '/goodbye', '/hello', '/'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(1)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re-render correctly when "push" action is called', () => {
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.push('/hello')
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
  })

  it('should re-render correctly when "goBack" action is called', () => {
    const history = createHistory({
      initialIndex: 1,
      initialEntries: ['/hello', '/goodbye'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.goBack()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re-render correctly when "go" action is called', () => {
    const history = createHistory({
      initialIndex: 2,
      initialEntries: ['/hello', '/goodbye', '/'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.go(-2)
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re-render correctly when "replace" action is called', () => {
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.replace('/hello')
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
  })

  it('should re-render correctly when "replace" action is called with routePath prop', () => {
    const history = createHistory({
      initialEntries: ['/hello/en'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route
                path="/hello/:language(fr|en)"
                routePath="/hello"
                component={HelloComponent}
              />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.replace('/hello/fr')
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello/:language(fr|en)',
          routeMatch: {
            url: '/hello/en',
            isExact: true,
            path: '/hello/:language(fr|en)',
            params: { language: 'en' },
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello/:language(fr|en)',
          routeMatch: {
            url: '/hello/fr',
            isExact: true,
            path: '/hello/:language(fr|en)',
            params: { language: 'fr' },
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
  })

  it('should re-render correctly when "onIndexChange" action is called with index arg', () => {
    let onIndexChange
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      onIndexChange = props.onIndexChange
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    onIndexChange(2)
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(3)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[2][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re-render correctly when "onIndexChange" action is called with index arg', () => {
    let onIndexChange
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      onIndexChange = props.onIndexChange
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    onIndexChange({ routeName: '/goodbye' })
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(3)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[2][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re render correctly when new tabs are provided', () => {
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    let component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    component.update(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={HelloComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            path: '/',
            isExact: true,
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[0][0].tabs).toMatchObject([
      {
        exact: true,
        path: '/',
        component: IndexComponent,
      },
      {
        path: '/hello',
        component: HelloComponent,
      },
      {
        path: '/goodbye',
        component: GoodbyeComponent,
      },
    ])
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            path: '/',
            isExact: true,
            params: {},
          },
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].tabs).toMatchObject([
      {
        exact: true,
        path: '/',
        component: HelloComponent,
      },
    ])
  })
})
