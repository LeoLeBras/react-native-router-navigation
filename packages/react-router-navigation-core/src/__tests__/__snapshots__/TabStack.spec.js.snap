// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabStack /> should re render correctly when new tabs are provided 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re render correctly when new tabs are provided 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "go" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "go" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "goBack" action is called 1`] = `
<Text>
  Goodbye
</Text>
`;

exports[`<TabStack /> should re-render correctly when "goBack" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "onIndexChange" action is called with index arg 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "onIndexChange" action is called with index arg 2`] = `
<Text>
  Goodbye
</Text>
`;

exports[`<TabStack /> should re-render correctly when "onIndexChange" action is called with index arg 3`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "onIndexChange" action is called with index arg 4`] = `
<Text>
  Goodbye
</Text>
`;

exports[`<TabStack /> should re-render correctly when "push" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "push" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called with routePath prop 1`] = `
<Text>
  {"language":"en"}
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called with routePath prop 2`] = `
<Text>
  {"language":"fr"}
</Text>
`;

exports[`<TabStack /> should render correctly 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should render correctly with initialIndex and initialEntries props  1`] = `
<Text>
  Hello
</Text>
`;
